snapshot_dbg_cli/set_logpoint_command.py

```diff
@@ -0,0 +1,208 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains the support for the set_logpoint command.
+
+The set_logpoint command is used to create a logpoint on a debug target
+(Debuggee).
+"""
+
+import argparse
+
+from enum import Enum
+
+from snapshot_dbg_cli import breakpoint_utils
+from snapshot_dbg_cli.exceptions import SilentlyExitError
+
+
+class LogLevel(Enum):
+  """ Enum that represents the logging level to use for emitted logpoints.
+
+  For storing the log level in the breakpoint data that the agents consume, we
+  use uppercase. When accepting the argument from the user in the set_logpoint
+  command we use lower case.
+  """
+
+  INFO = 'INFO'
+  WARNING = 'WARNING'
+  ERROR = 'ERROR'
+
+  def __str__(self):
+    return str(self.value)
+
+  @staticmethod
+  def parse_arg(log_level_arg):
+    # We use user friendly lowercase strings from the user.
+    mappings = {
+        'info': LogLevel.INFO,
+        'warning': LogLevel.WARNING,
+        'error': LogLevel.ERROR
+    }
+    enum_val = mappings.get(log_level_arg, None)
+
+    if enum_val is None:
+      raise argparse.ArgumentTypeError(
+          f'Invalid log-level argument provided: {log_level_arg}')
+
+    return enum_val
+
+
+DESCRIPTION = """
+Adds a debug logpoint to a debug target (debuggee). Logpoints inject logging
+into running services without changing your code or restarting your application.
+Every time any instance executes code at the logpoint location, Snapshot
+Debugger logs a message.  Output is sent to the standard log for the programming
+language of the target (java.logging for ava, logging for Python, etc.)
+
+Logpoints remain active for 24 hours after creation, or until they are deleted
+or the service is redeployed. If you place a logpoint on a line that receives
+lots of traffic, Debugger throttles the logpoint to reduce its impact on your
+application.
+"""
+
+LOCATION_HELP = """
+Specify the location to add the logpoint. Locations are of the form FILE:LINE,
+where FILE is the file name, or the file name preceded by enough path components
+to differentiate it from other files with the same name. If the file name isn't
+unique in the debuggee, the behavior is unspecified.
+"""
+
+LOG_FORMAT_STRING_HELP = """
+Specify a format string which will be logged every time the logpoint location is
+executed. If the string contains curly braces ('{' and '}'), any text within the
+curly braces will be interpreted as a run-time expression in the debug target's
+language, which will be evaluated when the logpoint is hit.
+
+The value of the expression will then replace the {} expression in the resulting
+log output. For example, if you specify the format string "a={a}, b={b}", and
+the logpoint is hit when local variable a is 1 and b is 2, the resulting log
+output would be "a=1, b=2".
```

Comment:
Probably worth mentioning that the braces can contain expressions.

Reply:
That's covered in the first part of the description, but I update the example to contain an expression to make it more obvious.